src/core/lib/transport/metadata_batch.cc

```diff
@@ -163,17 +172,28 @@ grpc_error* grpc_metadata_batch_link_head(grpc_metadata_batch* batch,
   return GRPC_ERROR_NONE;
 }
 
+grpc_error* grpc_metadata_batch_link_head(
```

Comment:
How do these changes handle the "Unallowed duplicate metadata" case that's handled in the maybe_link_callout case? Is it just through the GPR_DEBUG_ASSERT on line 100?

Reply:
My view is:
1) Our code should only be setting the relevant metadata exactly once,
2) If a user tries to set these metadata specifically, the existing behaviour is we give them a runtime error. IMO the way to do it is to give them a 'stop the world' error during their initial coding/debug phase so it is known that they should not be doing it.

Reply:
Just to confirm: Are the grpc_metadata_batch_callouts_index indices added only by our internal code? And there's no metadata we could receive that would cause us to add duplicates? In that case, I agree that we can rely on our code having the correct behavior.

Reply:
Hm...so I think there is a path where external callers could screw us up, so I've had it mirror the existing behaviour instead.

Reply:
Is it just one path? Is there a way to handle that logic elsewhere? I'm aware that having to revert your changes here has probably cost us some performance savings.

Reply:
The main issue comes from the user being able to add whatever metadata they want - one possibility would be to see if user added metadata necessarily happens after we add our own metadata, and in that case we can have a method for them that does the check, and an internal method that does not.

However, properly analyzing that will take some time and require review from @roth - so my thought is to put a TODO here for now.

Reply:
A TODO here would be great and/or following up with roth@.